cli/listener/src/main.rs

```diff
@@ -531,81 +694,168 @@ fn main() {
-    )
-    .expect("Failed when getting APY for period");
-    let growth_factor = apy.unwrap().growth_factor();
-    assert_eq!(growth_factor, 1.0005282698770684); //  Checked on WA, precision difference in the last digit.
+    // When computing the APY, we have to call `growth_factor` which divides two
+    // Rational numbers. Previously when dividing two rationals our implementation
+    // returned another Rational. In other words, for dividing `a/b` by `c/d`, we
+    // did `a*d/b*c`. In this case, `a*d` or `b*c` could overflow, we now return an
+    // `f64` instead of a Rational and avoid multiplying two large numbers that
+    // could overflow.
+    #[test]
+    fn test_rationals_do_not_overflow() {
+        let conn = Connection::open_in_memory().expect("Failed to open sqlite connection.");
+        create_db(&conn).unwrap();
+        let exchange_rate = ExchangeRate {
+            id: 0,
+            timestamp: chrono::Utc.ymd(2022, 01, 28).and_hms(11, 58, 39),
+            slot: 108837851,
+            epoch: 270,
+            pool: SOLIDO_ID.to_owned(),
+            price_lamports_numerator: 1936245653069130,
+            price_lamports_denominator: 1893971837707973,
+        };
+        insert_price(&conn, &exchange_rate).unwrap();
+
+        let exchange_rate = ExchangeRate {
+            id: 0,
+            timestamp: chrono::Utc.ymd(2022, 02, 28).and_hms(11, 58, 39),
+            slot: 118837851,
+            epoch: 275,
+            pool: SOLIDO_ID.to_owned(),
+            price_lamports_numerator: 1936245653069130,
+            price_lamports_denominator: 1892971837707973,
+        };
+        insert_price(&conn, &exchange_rate).unwrap();
+
+        let apy = get_interval_price_for_period(
+            conn.unchecked_transaction().unwrap(),
+            chrono::Utc.ymd(2020, 7, 7).and_hms(0, 0, 0),
+            chrono::Utc.ymd(2022, 7, 8).and_hms(0, 0, 0),
+            SOLIDO_ID.to_owned(),
+        )
+        .expect("Failed when getting APY for period");
+        let growth_factor = apy.unwrap().growth_factor();
+        assert_eq!(growth_factor, 1.0005282698770684); //  Checked on WA, precision difference in the last digit.
+    }
+
+    #[test]
+    fn test_get_date_from_url_parameters() {
+        let query_params = form_urlencoded::Serializer::new(String::new())
+            .append_pair("from", "2022-02-04T11:40:02.683960+00:00")
+            .append_pair("to", "2022-02-07T14:22:08.826526+00:00")
+            .finish();
+
+        let dates =
+            get_date_params(form_urlencoded::parse(query_params.as_bytes()).collect::<Vec<_>>());
+        assert_eq!(
+            dates,
+            Ok(parse_utc_iso8601("2022-02-04T11:40:02.683960+00:00")
+                ..parse_utc_iso8601("2022-02-07T14:22:08.826526+00:00")),
+        );
+    }
+
+    #[test]
+    fn test_get_interval_from_url() {
+        // check for correct and specific prices.
+        fn check_correct_url(conn: &Connection, url: &str) {
```

Comment:
It might be a bit nicer to split this into two tests, one that tests parsing of the url query params, and a different one that tests range selection. Then it’s easier to to see which part is broken when a test fails, and if all of the individual pieces pass the tests, we assume that the combination is good as well. (It is also good to test the full flow, but in that case I would try to test it externally, i.e. start the actual binary with some prepared database, make an http request to it, and verify the response. For our tiny server I think that’s overkill though, just the unit tests and some manual testing will do for now.)

Reply:
Why should we split this one? It only tests the successful queries. We can make another that tests for parsing errors.
We can also just delete this test, but I see no reason why, now that it is done 😬

Reply:
This test tests two things at once: parsing the query params, and running the query. We can test two smaller units independently. That helps to pinpoint a problem when one of them fails. But I don’t have a strong opinion on it.

Reply:
Yes, but parsing the query is tested already when testing `get_date_params`.